src/cli/subcommand/repair.rs

```diff
@@ -21,40 +36,538 @@ pub struct Repair {
+
         Ok(ExitCode::SUCCESS)
     }
 }
+
+#[derive(Serialize, Deserialize)]
+/// Structured output of `dscl -plist . -read /Groups/{name}`
+struct GroupPlist {
+    #[serde(rename = "dsAttrTypeStandard:GroupMembership")]
+    group_membership: Vec<String>,
+    #[serde(
+        rename = "dsAttrTypeStandard:PrimaryGroupID",
+        deserialize_with = "deserialize_gid"
+    )]
+    gid: u32,
+}
+
+pub fn deserialize_gid<'de, D>(deserializer: D) -> Result<u32, D::Error>
+where
+    D: serde::de::Deserializer<'de>,
+{
+    let s: Vec<String> = serde::Deserialize::deserialize(deserializer)?;
+
+    let gid_str = s
+        .first()
+        .ok_or_else(|| serde::de::Error::invalid_length(0, &"a gid entry"))?;
+
+    let gid: u32 = gid_str.parse().map_err(serde::de::Error::custom)?;
+
+    Ok(gid)
+}
+
+#[tracing::instrument]
+async fn get_existing_receipt() -> Option<InstallPlan> {
+    match std::path::Path::new(RECEIPT_LOCATION).exists() {
+        true => {
+            tracing::debug!("Reading existing receipt");
+            let install_plan_string = tokio::fs::read_to_string(RECEIPT_LOCATION).await.ok();
+
+            match install_plan_string {
+                Some(s) => match serde_json::from_str::<InstallPlan>(s.as_str()) {
+                    Ok(plan) => {
+                        tracing::debug!(plan_version = %plan.version, "Able to parse receipt");
+                        Some(plan)
+                    },
+                    Err(e) => {
+                        tracing::debug!(?e);
+                        tracing::warn!("Could not parse receipt. Your receipt will not be updated to account for the new UIDs");
+                        None
+                    },
+                },
+                _ => None,
+            }
+        },
+        false => None,
+    }
+}
+
+#[tracing::instrument(skip_all)]
+fn find_users_and_groups(
+    existing_receipt: Option<InstallPlan>,
+) -> color_eyre::Result<Option<(InstallPlan, usize, CreateUsersAndGroups)>> {
+    let ret = match existing_receipt {
+        Some(receipt) => {
+            tracing::debug!("Got existing receipt");
+
+            let mut maybe_create_users_and_groups_idx_action = None;
+            for (idx, stateful_action) in receipt.actions.iter().enumerate() {
+                let action_tag = stateful_action.inner_typetag_name();
+                tracing::trace!("Found {action_tag} in receipt");
+
+                if action_tag == CreateUsersAndGroups::action_tag().0 {
+                    tracing::debug!(
+                        "Found {} in receipt, preparing to roundtrip to extract the real type",
+                        CreateUsersAndGroups::action_tag().0
+                    );
+                    // NOTE(cole-h): this round-trip is kinda jank... but Action is not
+                    // object-safe, and I can't think of any other way to get the
+                    // concrete `CreateUsersAndGroups` type out of a `Box<dyn Action>`.
```

Comment:
fwiw, from [downstream review](https://gerrit.lix.systems/c/lix-installer/+/1912/2): https://doc.rust-lang.org/std/boxed/struct.Box.html#method.downcast-2

But this is not a big deal honestly, assuming that our roundtripping isn't busted.

Reply:
I tried that, but couldn't get it to work 🤷 

Maybe because it was originally a `Box<dyn Action>` and not `CreateUsersAndGroups` (i.e. it didn't go `CreateUsersAndGroups` -> `Box<dyn Action>`, but we parsed the receipt into a structure that only has `Vec<Box<dyn Action>>`, so it loses the original type information)?

If you can get it to work with that, I'd love to see the diff, because round-tripping isn't the most elegant... Maybe I overlooked something obvious!

Reply:
i think you need an Any bound, but the way to actually make all this work is unnecessarily arcane (i don't remember if it works as a supertrait or if you have to do it a different way). I'll have to consult what i did in clipper with dyn stuff and have a go if i remember to get around to it.